server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java

```diff
@@ -0,0 +1,411 @@
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new FutureThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = unionBlobStore.save(new ByteArrayInputStream(BLOB_CONTENT)).get();
+
+            SoftAssertions.assertSoftly(softly -> {
+                softly.assertThat(unionBlobStore.read(blobId))
+                    .hasSameContentAs(new ByteArrayInputStream(BLOB_CONTENT));
+                softly.assertThat(legacyBlobStore.read(blobId))
+                    .hasSameContentAs(new ByteArrayInputStream(BLOB_CONTENT));
+            });
+        }
+
+    }
+
+    @Nested
+    class CurrentReadThrowsExceptionDirectly {
+
+        @Test
+        void readShouldReturnFallbackToLegacyWhenCurrentGotException() throws Exception {
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new ThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = legacyBlobStore.save(BLOB_CONTENT).get();
+
+            assertThat(unionBlobStore.read(blobId))
+                .hasSameContentAs(new ByteArrayInputStream(BLOB_CONTENT));
+        }
+
+        @Test
+        void readBytesShouldReturnFallbackToLegacyWhenCurrentGotException() throws Exception {
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new ThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = legacyBlobStore.save(BLOB_CONTENT).get();
+
+            assertThat(unionBlobStore.readBytes(blobId).get())
+                .isEqualTo(BLOB_CONTENT);
+        }
+
+    }
+
+    @Nested
+    class CurrentReadCompletesExceptionally {
+
+        @Test
+        void readShouldReturnFallbackToLegacyWhenCurrentCompletedExceptionally() throws Exception {
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new FutureThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = legacyBlobStore.save(BLOB_CONTENT).get();
+
+            assertThat(unionBlobStore.read(blobId))
+                .hasSameContentAs(new ByteArrayInputStream(BLOB_CONTENT));
+        }
+
+        @Test
+        void readBytesShouldReturnFallbackToLegacyWhenCurrentCompletedExceptionally() throws Exception {
+            MemoryBlobStore legacyBlobStore = new MemoryBlobStore(BLOB_ID_FACTORY);
+            UnionBlobStore unionBlobStore = new UnionBlobStore(new FutureThrowingBlobStore(), legacyBlobStore);
+            BlobId blobId = legacyBlobStore.save(BLOB_CONTENT).get();
+
+            assertThat(unionBlobStore.readBytes(blobId).get())
+                .isEqualTo(BLOB_CONTENT);
+        }
+    }
+
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
```

Comment:
normally `@MethodSource` - parameter factory, only able to access static method. I need to declare `CurrentAndLegacyCouldNotComplete` test instance as `PER_CLASS`. This should help `@MethodSource` access to non static method factory. see [junit 5 documentaion](https://junit.org/junit5/docs/5.2.0/user-guide/#writing-tests-parameterized-tests-sources-MethodSource)